Firebase/Storage/FIRStorageReference.m

```diff
@@ -331,6 +333,91 @@ - (void)downloadURLWithCompletion:(FIRStorageVoidURLError)completion {
   [task enqueue];
 }
 
+#pragma mark - List
+
+- (void)listWithMaxResults:(int64_t)maxResults completion:(FIRStorageVoidListError)completion {
+  if (maxResults <= 0 || maxResults > 1000) {
+    [NSException raise:NSInvalidArgumentException
```

Comment:
I wonder, if the error is so bad that we want to raise an exception? Usually on iOS an exception means a hardly recoverable error when it is better crash the app rather than continue. It seams to me that here we can just pass an error to the completion.

Reply:
This is what we do on Firestore (there is no API validation in Storage yet). We terminate for invalid API usage, and use the completion for errors that are harder to prevent.

Reply:
I am still not sure if it is a good approach for Objective-C. Objective-C exceptions and their handling are much more limited comparing to C++ exceptions, so it is better to avoid using them until it is necessary.

Just an opinion: as a client developer I would hate an Objective-C lib that throws exception just because of a requested `maxResults` is e.g. 1 item more than the lib can give (especially if there is a callback where the error can be passed)

Reply:
My time on Firestore has left me very biased here. I'll let Ryan Wilson chime in.

Reply:
I would say in this case it would make sense to return an error instead, or automatically cap it at 1000. I think it'd be pretty easy for a naive implementation to get the number from a different data source (say, a value from Firestore) and passes in a count there.

If the developer builds this appropriately, they'll use the returned page token so they shouldn't be using hardcoded indexes anyway so modifying the `maxPageSize` should be acceptable.

Thoughts? 😄

Reply:
Modifying of `maxPageSize` here sounds good to me

Reply:
I really don't want to modify the `maxPageSize`. A lot of people will write code that assumes that there are no more results if `items.count < maxPageSize`. While that is not guaranteed to begin with, it's usually true. 

I will update the PR to invoke the error callback and hope to race it into the master branch before Ryan can fix the release tooling :)

Reply:
PR updated. 

I do want to point out https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/ErrorHandling/ErrorHandling.html:

```
When you’re writing code with Objective-C, exceptions are used solely for programmer errors, like out-of-bounds array access or invalid method arguments. These are the problems that you should find and fix during testing before you ship your app.
```

But either way, we are invoking the callback now with an NSError.